### ISSUE TITLE
dsnt in testing phase

### ISSUE THREAD

**Issue by IQ17** [anibali/dsntnn#13]
Thanks for the paper and code!

I am doing pose estimation and face a problem that the heatmap for predicting the left wrist also fires a little response on the right wrist, which means, the heatmap has two peaks, a strong peak on the left wrist and a weak peak on the right wrist. 

The two peaks problem makes the dsnt predicted result uncorrect, do you have any suggestions? Thanks!

**Comment by IQ17** [comment]
I understand the argument in paper that two peaks heatmap should be punished by the loss, while in the testing phase, ambiguity does happen sometimes.

**Comment by anibali** [comment]
As you mention in your second comment, the overall idea is that the model simply learns not to generate multiple peaks, and I have found this to mostly be a good assumption---it is relatively easy for the model to use ReLU thresholding to suppress smaller peaks.

However, if you really want to, there's nothing to stop you from manipulating the heatmap yourself at test time. You could, for example, apply your own thresholding or filtering before applying DSNT. You could even use argmax to find the brightest pixel, darken areas of the heatmap that are really far from that position, and then apply DSNT. Since you don't need to be able to backprop at test time, the sky's the limit!

**Comment by IQ17** [comment]
Thanks for the reply! 
My raw_heatmaps is obtained as following:
feature maps from backbone network -> raw_heatmaps by applying 16 1x1 conv -> batch norm and relu6 -> dsnt

**Comment by anibali** [comment]
That sounds strange, training with only 4 human pose images should be able to overfit to heatmaps which are very close to ground truth. It sounds like you have a problem with your specific setup. I can say with fairly high confidence that DSNT works for human pose estimation as I've used it for that purpose many times before ;)

1. Are there any weighted layers (ie. fully connected or convolutional) after that last batch norm you mentioned? If not, you should probably remove it since it might be restricting the kinds of heatmaps that can be generated by the network.
2. Are you training the parameters of the backbone network as well as the 1x1 conv? If not, that single layer may not have enough capacity to transform your features into good heatmaps.
3. Are you normalising the heatmaps with `dsntnn.flat_softmax` or similar?

**Comment by IQ17** [comment]
Thanks for the instructions!

1. There is no weighted layers after the batch norm, so I removed the batch norm as you suggested but kept the relu6, so now it looks like:
feature maps from backbone network -> raw_heatmaps by applying 16 1x1 conv ->  relu6 -> dsnt

2. I train the whole network, including the backbone

3. Actually I am in tensorflow, and thus use https://github.com/ashwhall/dsnt , which has softmax normalization and others, and I used softmax normalization.

Some results below. 
I trained a single human pose image, and printed out the coordinates and showed the normalized heatmaps.

ground truth (x, y) for joints
[[ 0.06770834 -0.578125  ]
 [ 0.078125   -0.28645834]
 [-0.06770834 -0.203125  ]
 [-0.08854166 -0.00520833]
 [-0.09895834  0.171875  ]
 [ 0.22395833 -0.27604166]
 [ 0.171875   -0.140625  ]
 [ 0.28645834  0.046875  ]
 [-0.24479167  0.109375  ]
 [-0.078125    0.328125  ]
 [-0.05729167  0.8697917 ]
 [ 0.00520833  0.140625  ]
 [ 0.015625    0.5364583 ]
 [-0.234375    0.328125  ]]

predicted (x, y) for joints, most of them are indeed close to the ground truth
[[ 0.0677143  -0.5781198 ]
 [ 0.07813136 -0.2864602 ]
 [-0.06770479 -0.20312512]
 [-0.02403468 -0.00092831]
 [-0.09896109  0.16489528]
 [ 0.2239613  -0.27604097]
 [ 0.17187695 -0.14062575]
 [ 0.15510873 -0.10488734]
 [-0.24478546  0.10938045]
 [-0.14869449  0.13047174]
 [-0.18530153  0.56645685]
 [-0.04927395  0.05036128]
 [-0.0202063   0.38146675]
 [-0.23438299  0.32812387]]

**Comment by IQ17** [comment]
Thanks for the instructions!

1. There is no weighted layers after the batch norm, so I removed the batch norm as you suggested but kept the relu6, so now it looks like:
feature maps from backbone network -> raw_heatmaps by applying 16 1x1 conv -> relu6 -> dsnt

2. I train the whole network, including the backbone

3. Actually I am in tensorflow, and thus use https://github.com/ashwhall/dsnt , which has softmax normalization and others, and I used softmax normalization. Also I use the average Jensen-Shannon divergence as regularization (default in code)

Some results below.
I trained a single human pose image, and printed out the coordinates and showed the normalized heatmaps.

ground truth (x, y) for joints
[[ 0.06770834 -0.578125 ]
[ 0.078125 -0.28645834]
[-0.06770834 -0.203125 ]
[-0.08854166 -0.00520833]
[-0.09895834 0.171875 ]
[ 0.22395833 -0.27604166]
[ 0.171875 -0.140625 ]
[ 0.28645834 0.046875 ]
[-0.24479167 0.109375 ]
[-0.078125 0.328125 ]
[-0.05729167 0.8697917 ]
[ 0.00520833 0.140625 ]
[ 0.015625 0.5364583 ]
[-0.234375 0.328125 ]]

predicted (x, y) for joints, most of them are indeed close to the ground truth
[[ 0.0677143 -0.5781198 ]
[ 0.07813136 -0.2864602 ]
[-0.06770479 -0.20312512]
[-0.02403468 -0.00092831]
[-0.09896109 0.16489528]
[ 0.2239613 -0.27604097]
[ 0.17187695 -0.14062575]
[ 0.15510873 -0.10488734]
[-0.24478546 0.10938045]
[-0.14869449 0.13047174]
[-0.18530153 0.56645685]
[-0.04927395 0.05036128]
[-0.0202063 0.38146675]
[-0.23438299 0.32812387]]

the drawed image
![image](https://user-images.githubusercontent.com/5336719/57007158-214a2400-6c19-11e9-99c2-9de66083de0c.png)

normalized heatmaps, most of them are multi-peaks, but the (x, y) expectation are right (verified by numpy calculation)

![pred_normalised_heatmaps_val0](https://user-images.githubusercontent.com/5336719/57007207-8736ab80-6c19-11e9-989a-cf300503b547.jpg)
![pred_normalised_heatmaps_val1](https://user-images.githubusercontent.com/5336719/57007208-8a319c00-6c19-11e9-93eb-6ca69b209c4b.jpg)
![pred_normalised_heatmaps_val2](https://user-images.githubusercontent.com/5336719/57007209-8d2c8c80-6c19-11e9-9d22-551493f41752.jpg)
![pred_normalised_heatmaps_val3](https://user-images.githubusercontent.com/5336719/57007217-9584c780-6c19-11e9-874d-4cfbd8595814.jpg)
![pred_normalised_heatmaps_val4](https://user-images.githubusercontent.com/5336719/57007219-99b0e500-6c19-11e9-907e-d3957698fdac.jpg)
![pred_normalised_heatmaps_val5](https://user-images.githubusercontent.com/5336719/57007220-9c133f00-6c19-11e9-81f7-f46f9796d4b3.jpg)
![pred_normalised_heatmaps_val6](https://user-images.githubusercontent.com/5336719/57007224-9e759900-6c19-11e9-94f8-92bce9304519.jpg)
![pred_normalised_heatmaps_val7](https://user-images.githubusercontent.com/5336719/57007228-a2a1b680-6c19-11e9-83f7-c9f58fbdec00.jpg)
![pred_normalised_heatmaps_val8](https://user-images.githubusercontent.com/5336719/57007231-a59ca700-6c19-11e9-8cb4-b763c38beb70.jpg)
![pred_normalised_heatmaps_val9](https://user-images.githubusercontent.com/5336719/57007232-a7666a80-6c19-11e9-8dd9-0cf20a4e8083.jpg)
![pred_normalised_heatmaps_val10](https://user-images.githubusercontent.com/5336719/57007233-a9c8c480-6c19-11e9-8521-8625a3b05d1e.jpg)
![pred_normalised_heatmaps_val11](https://user-images.githubusercontent.com/5336719/57007235-acc3b500-6c19-11e9-97e5-af9e37ddbe61.jpg)
![pred_normalised_heatmaps_val12](https://user-images.githubusercontent.com/5336719/57007236-af260f00-6c19-11e9-89b7-a01f27bfd15a.jpg)
![pred_normalised_heatmaps_val13](https://user-images.githubusercontent.com/5336719/57007238-b1886900-6c19-11e9-81c7-93e16029072a.jpg)

**Comment by anibali** [comment]
Ah, if the predicted points are mostly correct then that changes things a bit. Playing with the regularisation coefficient like you suggest might help, since from the perspective of the Euclidean loss everything is going fine. You might also find that the problem simply goes away with more training examples, since learning these "degenerate" heatmaps is probably not especially generalisable.

Also, I'm not sure why you have that last relu6 in there either. Remember, the softmax normalisation _is_ an activation function. In theory I think that the network should be able to cope with having the relu there as well, but I'm not sure if it's a good idea.

**Comment by IQ17** [comment]
Actually, by looking into the [code](https://github.com/ashwhall/dsnt/blob/149a1db248635be926fbe305c14018cceec89cdb/dsnt.py#L112) , I found out that it assumes the (x, y) coordinates are within [0, 1], instead of [-1, 1], so all my ground truth gaussian heatmaps are wrong....sadly. By fixing this bug, my regularisation loss reduces to 0.0004343366 quickly in my 4 images overfitting test, and all heatmaps are single peaks, beautifully!

Regarding the relu6, you are right, it is unnecessary or even a little harmful. I put it there just hope it could suppress some peaks in heatmaps. Since the multi-peaks problem is caused by a bug, I remove the relu6.

Now I will train the whole dataset. 
Really appreciate your kindly suggestions and wonderful paper!

**Comment by anibali** [comment]
Great to hear that you figured it out! I'll add a warning to my README about the Tensorflow implementation using different coordinates.
